File: api/app/models/actuals.py
"""Actuals models - time tracking and signing."""
import uuid
from datetime import datetime
from typing import Optional
from sqlalchemy import (
    String,
    Integer,
    DateTime,
    Text,
    ForeignKey,
    Index,
    CheckConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from api.app.db.base import Base


def generate_uuid() -> str:
    return str(uuid.uuid4())


class ActualLine(Base):
    """Actual line - recorded time spent by resource on project."""
    __tablename__ = "actual_lines"
    
    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=generate_uuid)
    tenant_id: Mapped[str] = mapped_column(String(36), nullable=False, index=True)
    period_id: Mapped[str] = mapped_column(String(36), ForeignKey("periods.id"), nullable=False)
    resource_id: Mapped[str] = mapped_column(String(36), ForeignKey("resources.id"), nullable=False)
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id"), nullable=False)
    year: Mapped[int] = mapped_column(Integer, nullable=False)
    month: Mapped[int] = mapped_column(Integer, nullable=False)
    
    # FTE values
    planned_fte_percent: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)  # From demand
    actual_fte_percent: Mapped[int] = mapped_column(Integer, nullable=False)
    
    # Employee signature
    employee_signed_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    employee_signed_by: Mapped[Optional[str]] = mapped_column(String(36), nullable=True)  # Could be proxy
    is_proxy_signed: Mapped[bool] = mapped_column(Integer, default=False)  # SQLite-friendly bool
    proxy_sign_reason: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    # RO approval
    ro_approved_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    ro_approved_by: Mapped[Optional[str]] = mapped_column(String(36), nullable=True)
    
    # Metadata
    created_by: Mapped[str] = mapped_column(String(36), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    resource: Mapped["Resource"] = relationship("Resource")
    project: Mapped["Project"] = relationship("Project")
    
    __table_args__ = (
        # FTE must be between 5 and 100 (or 0 for no work)
        CheckConstraint(
            '(actual_fte_percent = 0) OR (actual_fte_percent >= 5 AND actual_fte_percent <= 100)',
            name='ck_actual_fte_range'
        ),
        # FTE must be divisible by 5
        CheckConstraint('actual_fte_percent % 5 = 0', name='ck_actual_fte_step'),
        # Unique per resource per project per month
        Index(
            'ix_actual_unique',
            'tenant_id', 'resource_id', 'project_id', 'year', 'month',
            unique=True,
        ),
        Index('ix_actual_tenant_period', 'tenant_id', 'year', 'month'),
        Index('ix_actual_resource', 'tenant_id', 'resource_id', 'year', 'month'),
    )
    
    @property
    def is_signed(self) -> bool:
        """Check if the actual line is signed."""
        return self.employee_signed_at is not None


# Import models used in relationships
from api.app.models.core import Resource, Project

File: frontend/src/api/actuals.ts
/**
 * Actuals API calls
 */
import { apiClient } from './client';

export interface ActualLine {
  id: string;
  tenant_id: string;
  period_id: string;
  resource_id: string;
  project_id: string;
  year: number;
  month: number;
  planned_fte_percent: number | null;
  actual_fte_percent: number;
  employee_signed_at?: string;
  employee_signed_by?: string;
  is_proxy_signed: boolean;
  proxy_sign_reason?: string;
  ro_approved_at?: string;
  ro_approved_by?: string;
  created_by: string;
  created_at: string;
}

export interface CreateActualLine {
  period_id: string;
  resource_id: string;
  project_id: string;
  year: number;
  month: number;
  planned_fte_percent?: number | null;
  actual_fte_percent: number;
}

export interface ResourceMonthlyTotal {
  resource_id: string;
  year: number;
  month: number;
  total_actual_fte: number;
  remaining: number;
  lines: ActualLine[];
}

export const actualsApi = {
  async getActualLines(periodId?: string, year?: number, month?: number): Promise<ActualLine[]> {
    const params = new URLSearchParams();
    if (periodId) params.append('period_id', periodId);
    if (year) params.append('year', String(year));
    if (month) params.append('month', String(month));
    const query = params.toString() ? `?${params.toString()}` : '';
    return apiClient.get<ActualLine[]>(`/actuals${query}`);
  },
  
  async getMyActuals(year?: number, month?: number): Promise<ActualLine[]> {
    const params = new URLSearchParams();
    if (year) params.append('year', String(year));
    if (month) params.append('month', String(month));
    const query = params.toString() ? `?${params.toString()}` : '';
    return apiClient.get<ActualLine[]>(`/actuals/my${query}`);
  },
  
  async getMyResource(): Promise<{ resource_id: string | null }> {
    return apiClient.get<{ resource_id: string | null }>('/actuals/my-resource');
  },
  
  async getResourceMonthlyTotal(resourceId: string, year: number, month: number): Promise<ResourceMonthlyTotal> {
    return apiClient.get<ResourceMonthlyTotal>(`/actuals/resource/${resourceId}/month?year=${year}&month=${month}`);
  },
  
  async createActualLine(data: CreateActualLine): Promise<ActualLine> {
    return apiClient.post<ActualLine>('/actuals', data);
  },
  
  async updateActualLine(id: string, data: Partial<CreateActualLine>): Promise<ActualLine> {
    return apiClient.put<ActualLine>(`/actuals/${id}`, data);
  },
  
  async deleteActualLine(id: string): Promise<void> {
    return apiClient.delete(`/actuals/${id}`);
  },
  
  async signActuals(actualId: string): Promise<ActualLine> {
    return apiClient.post<ActualLine>(`/actuals/${actualId}/sign`, {});
  },
  
  async proxySignActuals(actualId: string, reason: string): Promise<ActualLine> {
    return apiClient.post<ActualLine>(`/actuals/${actualId}/proxy-sign`, { reason });
  },
};
